### ISSUE TITLE
"StrictFilePathMarshaler must be used on a FilePath" error

### ISSUE THREAD

**Issue by deadlydog** [deadlydog/VS.DiffAllFiles#14]
The following error was reported on the VS Marketplace:

> Trying to use this. When I select a changed file and click Unstaged (or All Files), I get: 
> Unexpected Error Occurred:
> StrictFilePathMarshaler must be used on a FilePath.
> V15.2(26430.16) Release.
> Diff All Files v1.0.1
> By dale 8/8/2017

I've experienced this problem in the VS 2015 version of the Diff All Files, but not the VS 2017 version, so it will still need to be investigated. In VS 2015 this error was thrown from the gitlib2sharp library, and the workaround was to re-clone the repository. I'm not sure if that workaround will work in VS 2017 or not though, so if we could get confirmation on that it would be nice.

We still need to see if there's some way we can have Diff All Files handle this error more gracefully, and ideally have additional logic in code to get it working without the user having to re-clone the Git repository.

**Comment by martinskuta** [comment]
I am encountering the same error in VS 2017 (15.3).

![image](https://user-images.githubusercontent.com/14070352/29874976-b665d28e-8d98-11e7-9f78-6dafdb710af9.png)

I tried to re-clone the repo as I read in your first comment, but it is still same error. Is there something else I could try?

**Comment by deadlydog** [comment]
Hi @martinskuta, are you still able to reproduce the error? If so, can you give this updated version 1.0.1.1 of Diff All Files a try and see if it resolves the issue. All I've done here is updated to use the latest version of LibGit2Sharp, and the release notes did mention some bug fixes around file paths in the library, so I'm hoping it solves the issue. Let me know, thanks.

[VS.DiffAllFiles.VS2017-v1.0.1.1.zip](https://github.com/deadlydog/VS.DiffAllFiles/files/1294933/VS.DiffAllFiles.VS2017-v1.0.1.1.zip)

Oh, and you'll need to rename the file extension from `.zip` to `.vsix` in order to install it; GitHub wouldn't allow me to upload it with the `.vsix` extension.

**Comment by martinskuta** [comment]
Hi,

I tried the attached version, it seems to solve the issue, but I get different error now:
![image](https://user-images.githubusercontent.com/14070352/30313502-e176291a-979d-11e7-8f0c-cfc2aaf3bdec.png)
It happens if I try with All files, staged, unstaged, combined or not. This unexpected error happens for every file (the notification popup there is given) Any idea?

**Comment by martinskuta** [comment]
Ah, now I noticed you have created #15 for it :)

**Comment by deadlydog** [comment]
Thanks for the confirmation @martinskuta. I'm glad LibGit2Sharp was able to get at least one of their problems resolved. I tried looking into the `key not present in the dictionary` error last night without much success. [This is where the error is thrown from](https://github.com/deadlydog/VS.DiffAllFiles/blob/master/VS.DiffAllFiles/GitHelper.cs#L156). I'm going to keep it on my mind though in hopes of having a eureka moment of how to get around it. If you have any suggestions, I'm all ears. Thanks.

**Comment by deadlydog** [comment]
I just updated the VS 2017 extension to v1.0.2 which includes the latest version of LibGit2Sharp (v0.26). I had a git repo where I could reliably reproduce the "The given key was not present in the dictionary." error, and it looks to have fixed it. I'm not certain if it also fixes the "StrictFilePathMarshaler must be used on a FilePath" error however as I do not have a git repo where I can reliably reproduce it anymore. If you guys do, I'd appreciate you testing it out and letting me know if it seems to be fixed with the new version or not. Thanks!

**Comment by ErikvO** [comment]
We just started using git and VS2019. I now get the "StrictFilePathMarshaler must be used on a FilePath" error in VS2019:
![image](https://user-images.githubusercontent.com/11665846/56812796-e40e1c80-683b-11e9-91e2-9a996f81dfd9.png)
Could there be a regression?

**Comment by deadlydog** [comment]
Thanks @ErikvO , it seems this error is unrelated to the "key not present in dictionary error" I guess. Would you be willing to zip up and send me your repository so that I can reproduce this issue and try and debug it? If so, send it to my email address deadlydog AT hotmail.com. Thanks!

**Comment by ErikvO** [comment]
Unfortunately I cannot share the repository. 
I have however found out what the issue was for me: I also had the GitHub extension installed. This extension also has a dependency on libgit2sharp.
[This](https://github.com/libgit2/libgit2sharp/issues/241) reported issue on the libgit2sharp repository pointed me into the right direction. I assume the versions of libgit2sharp used by GitHub and DiffAllFiles are different, causing this issue.

**Comment by deadlydog** [comment]
Thanks @ErikvO . I've found that same reported issue in the past, and is actually what lead me down the road of self-signing the libgit2sharp.dll that gets included with DiffAllFiles to prevent Visual Studio from mixing up the same dll from the 2 different sources. I've had the `GitHub for Visual Studio` extension installed in VS 2017 for years and now also have it in 2019, and since implementing the self-signing of the dll have not experienced this issue.

You're saying that you removed the `GitHub for Visual Studio` extension and the problem went away?

**Comment by ErikvO** [comment]
Yes, I indeed removed the `GitHub for Visual Studio` extension. That resolved the issue.
For now this is acceptable, as I don’t need it for my job. I just installed it so I could also work on my hobby projects at work.

**Comment by deadlydog** [comment]
@ErikvO I'm also curious if you saw that error on ALL of your git repos when you had the `GitHub for Visual Studio` extension installed, or only in some of them?

**Comment by ErikvO** [comment]
I saw the error in all of them. We use multiple repositories and they all had it.
I've removed a local repository and recloned it in another location. The issue also happened in this new repository.

**Comment by maxisam** [comment]
I have this issue on vs2019 with `GitHub for Visual Studio`

**Comment by deadlydog** [comment]
> I have this issue on vs2019 with `GitHub for Visual Studio`

@maxisam I'm not entirely surprised, as my extension and that one are both using the LibGit2Sharp library which is where this error is originating from.

I'm curious, are you getting that error in `GitHub for Visual Studio` even without having `Diff All Files` installed? Or do you have them both installed?

**Comment by maxisam** [comment]
it was fine with `GitHub for Visual Studio` until I installed `Diff All Files`

**Comment by wiz0u** [comment]
> I have this issue on vs2019 with `GitHub for Visual Studio`

I have this issue on vs2019 **without** `GitHub for Visual Studio`

I just made my first 'publish' of my project using VS2019 'Publish to Github' provider, and Diff All Files give me the same error as OP

(I don't have the problem with other projects that are using a normal git repository - non github)
